### PR TITLE
DDF-4478 Improves filtering logic for dropdowns

### PR DESCRIPTION
**Pull request by vinamartin**
#### What does this PR do?
Improves filtering logic for dropdowns in Intrigue.
Previously, filtering through a dropdown would match if any part of the filter was found in the word. This would lead to weird behavior like returning "validation" and "width" when the filter was "id".

The new behavior splits the words on camelcase, dashes, dots, and spaces then checks to see if any of the resulting words start with the filter value. This results in a slightly more intuitive behavior as shown here:
betterdropdown1
betterdropdown2

#### Who is reviewing it? 
@emmberk @ethantmanns @Kjames5269 @jhunzik 
@TeresaHudson (**UX**)
@andrewkfiedler 

#### Ask 2 committers to review/merge the PR and tag them here.
@AzGoalie
@djblue

#### How should this be tested?
Try using a few of the dropdowns in Intrigue as found in Advanced Search, Upload Editor, Metacard editing, etc.
I'm not sure if I chose good delimiters here so make sure to test with as many dropdowns as possible.

#### What are the relevant tickets?
[DDF-4478 Improve filtering logic for Advanced Search](https://codice.atlassian.net/browse/DDF-4478) 
Previous Behavior: 
Previous Behavior

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by djblue** [comment]
I think we might want to pull out the filter logic as a separate file so we can add some unit tests and reuse the logic without marionette.

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/select.collection.view.js [hunk 1]
```diff
@@ -170,6 +170,18 @@ module.exports = Marionette.CollectionView.extend({
     str = str.toString()
     return this.options.matchcase === true ? str : str.toLowerCase()
   },
+  getWords: function(str) {
+    //Handle camelcase
+    str = str.replace(/([A-Z])/g, ' $1')
```
Comment (ethantmanns) [inline ui/packages/catalog-ui-search/src/main/webapp/component/select/select.collection.view.js]: This goes against the conventions for our taxonomy names, but would this cause an attribute named `EXAMPLE` to not appear as a result when I user typed in `EXA`?

Comment (vinamartin) [reply ui/packages/catalog-ui-search/src/main/webapp/component/select/select.collection.view.js]: This is a good catch! Since users can actually input their own text for some text boxes this will break for that case.
screen shot 2019-01-28 at 10 15 04
screen shot 2019-01-28 at 10 15 13

**Comment by vinamartin** [comment]
Something else that needs to be addressed is filtering on a whole name like, "example.attribute-with-this-format" won't match once I get past "example."

**Comment by vinamartin** [comment]
It looks like regex might have been the way to go after all, this is the functionality I have now which seems to cover the corner cases:
betterdropdown3
screen shot 2019-01-28 at 16 18 34

I'm going to test and push up a fix.

**Comment by vinamartin** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+var filterHelper = require ('./filterHelper.js')
+
+describe('fitler attrs', () => {  
+    it('filter anyText', () => {
+      var str = "anyText"
+      var matchcase = false
+
+      var filterValue = "w"
+      expect(filterHelper.matchesFilter(filterValue, str, matchcase)).to.equal(false)
```
Comment (djblue) [inline ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js]: I would combine some of these test cases into a few big data structure to loop through and reduce duplication.

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+/*global define*/
+define([], function() {
```
Comment (djblue) [inline ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.js]: We are trying to consolidate the module formats we use. I eliminated amd (`define`/`return`) and prefer to write new code with es6 (`import`/`export`)

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js (hunk 2 above)
Comment (vinamartin) [reply ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js]: That's what I was thinking too, just wanted to get something out there 👍

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.js (hunk 3 above)
Comment (vinamartin) [reply ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.js]: Oh okay neat! I'll update this!

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+import { expect } from 'chai'
+
+var filterHelper = require ('./filterHelper.js')
+
+describe('fitler attrs', () => {
```
Comment (ethantmanns) [inline ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js]: ✏️: "fitler"

Comment (vinamartin) [reply ui/packages/catalog-ui-search/src/main/webapp/component/select/filterHelper.spec.js]: Nice catch!

**Comment by vinamartin** [comment]
build now

**Comment by vinamartin** [comment]
One thing that should still be solved on this is that hitting enter/clicking outside the dropdown doesn't let you select exactly what you want. 
For instance: I type in "id" and I know that's what I want. I still have to scroll.

File: ui/packages/catalog-ui-search/src/main/webapp/component/select/select.collection.view.js [hunk 5]
```diff
@@ -20,6 +20,7 @@ const $ = require('jquery')
 const childView = require('./select.view')
 const CustomElements = require('../../js/CustomElements.js')
 const Common = require('../../js/Common.js')
+import { matchesFilter, getAppropriateString } from './filterHelper'
```
Comment (emmberk) [inline ui/packages/catalog-ui-search/src/main/webapp/component/select/select.collection.view.js]: 🤔I think this is big improvement from before, but I think there is still some usability issues , especially when there are many attribute names that contain delimited words that start with the same letters.

My use case is searching by metacard id. Even though this ticket reduced the number of options that I see when I type "id", I still must scroll down the list to select "id", because there are many options that come before "id" alphabetically that contain the word "id" or that start with "id". I do think that it is user-friendly to alphabetize the options in the drop-down, but there could be an easier way to allow a user to search by an attribute by typing in the full name of the attribute.

**Comment by vinamartin** [comment]
I think we're going to take this as an initial solution and iterate on it.
build now
